File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.py
def solution(participant, completion):
    # 참가자 이름과 그 수를 딕셔너리(딕셔너리는 {})
    dict = {}

    for i in participant:
        if i in dict:
            dict[i] += 1
        else:
            dict[i] = 1

    # 완주자 목록을 순회하면서 참가자 카운트에서 뺍니다.
    for i in completion:
        dict[i] -= 1
        # 카운트가 0이 되면 딕셔너리에서 제거합니다.(이걸 생각)
        if dict[i] == 0:
            del dict[i]

    # 남아있는 참가자가 완주하지 못한 선수입니다.
    for i in dict:
        return i

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.HashMap;

class Solution {
    public String solution(String[] participant, String[] completion) {
        // 해시맵 생성
        HashMap<String, Integer> hashMap = new HashMap<>();

        // 참가자가 아닌 완주자들의 이름을 해시맵에 저장
        for (String string : completion){
            // getOrDefault(string,0) : hashMap 에서 키가 string 인 것을 가져옵니다.
            // 만약 string 이라는 키가 존재한다면 그 키에 대응
            // 만약 string 이라는 키가 존재하지 않으면 기본값으로 0 을 반환
            hashMap.put(string, hashMap.getOrDefault(string,0)+1);
        }

        // 참가한 선수들의 이름을 키로 하는 값을 1씩 감소
        for (String string : participant){
            // 해당 선수 이름이 해시맵에 없거나(완주자 명단에 없는 경우), 이미 다 처리된 경우(카운트가 0인 경우)
            if (hashMap.getOrDefault(string, 0) == 0){
                // 완주하지 못한 선수이므로 그 선수 이름을 반환
                return string;
            }
            // 해당 선수 이름의 카운트를 1 감소 (즉, 이 선수를 처리함)
            hashMap.put(string, hashMap.get(string)-1);
        }
        // 모든 선수가 완주한 경우 null 을 반환
        return null;
    }
}